### ISSUE TITLE
elgg_group_gatekeeper does not forward to login page when he should

### ISSUE THREAD

**Issue by oseg** [Elgg/Elgg#12048]
Context: elgg v2.3.7

Scenario N°1:  with an open public group whose 'accessibility of group content' is set to 'members only'
- A user which is member of the group, who has been logged out after long session for instance, try to access to the group bookmarks list url for instance
=> The following error message is displayed "You must be logged in to view the requested page", but the user is not forward to the login page

Scenario N°2: same scenario as N°1 but this time the group is not public, he is invisible
- A user which is member of the group, who has been logged out after long session for instance, try to access to the group bookmarks list url for instance
=> The following error message is displayed "The content you were trying to view has been removed or you do not have permission to view it"
=> From a user experience point of view, it would have been better to ask the user to login in a 1st step and then in a 2nd step display an error message or not according to the situation

**Comment by oseg** [comment]
Suggested modifications in elgg_group_gatekeeper()

```
function elgg_group_gatekeeper($forward = true, $group_guid = null) {
	if (null === $group_guid) {
		$group_guid = elgg_get_page_owner_guid();
	}

	if (!$group_guid) {
		return true;
	}

	// this handles non-groups and invisible groups
	$visibility = \Elgg\GroupItemVisibility::factory($group_guid);

	if (!$visibility->shouldHideItems) {
		return true;
	}
	if ($forward) {
                $reason_hidden = $visibility->reasonHidden;
                $forward_reason = 'system';
                $forward_url = '';
                switch($visibility->reasonHidden) {
                    default:
                    case 'no_access':
                        if (elgg_is_logged_in()){
                            //user is logged in but has no access
                        } else {
                            //Some content are only accessible to logged in users, please login
                            $reason_hidden = 'logged_out';
                            _elgg_services()->redirects->setLastForwardFrom();
                            $forward_reason = 'login';
                            $forward_url = '/login';
                        }
                        break;
                    case 'logged_out':
                        //forward to login page
			_elgg_services()->redirects->setLastForwardFrom();
			$forward_reason = 'login';
                        $forward_url = '/login';
                        break;
                    case 'non_member':
                        //forward to group page
			$forward_reason = 'member';
                        $group = get_entity($group_guid);
                        $forward_url = $group ? $group->getURL() : '';
                        break;
                }

		$msg_keys = array(
			'non_member' => 'membershiprequired',
			'logged_out' => 'loggedinrequired',
			'no_access' => 'noaccess',
		);
		register_error(elgg_echo($msg_keys[$reason_hidden]));
		forward($forward_url, $forward_reason);
	}

	return false;
}
```

**Comment by jeabakker** [comment]
Because the gatekeeper logic in Elgg 3 has changed (and the group gatekeeper has been replaced) trying to fix this in Elgg 2.3 and than have a different behavior in Elgg 3 is not worth the effort.
